### PR TITLE
feat: detail panel animation

### PR DESCRIPTION
**Pull request by Domino987**
## Related Issue
#258

## Description

This PR extracts the detail panel to its own component. Using refs and timeout ensures that the detail panel animation work for both the enter and exit animation.

Also only the current opened panel will be rendered as visible in the console

There is a breaking change because the detail panel is now a full react component.
```
render: ({ rowData }) => {...}
detailPanel: ({ rowData }) => {...}
```

To render a detailpanel.

out

### CODE REVIEW

**Comment by appukuttan-shailesh** [comment]
@Domino987 thanks for looking into this. I was working on this probably at the same time as you (See #261).
I like your implementation and was testing it out. It works great, except that it re-renders couple of times on exit (this can cause unnecessary execution of functions, as was happening in my original use case).
To see this easily, it is useful to add a some log statements to the demo app ([like here](https://github.com/appukuttan-shailesh/core/commit/e4e3aefbf08d7b8003723da225c5bf525883624b) and in #261).

Once you add that to your PR here, you will see this:
Alternative

(notice the console logs).

#261 currently works as below:
Peek 2021-07-16 13-53

**Comment by appukuttan-shailesh** [comment]
@Domino987 .... I was testing #260 a bit more and seem to find behavior different (erroneous) from that in your screen recording when trying on my pc.
See below:
Problem

To double-check I again cloned your branch and tested without making any changes (`npm install` followed by `npm run`). Wonder why we see different behaviors?

**Comment by Domino987** [comment]
Hi @appukuttan-shailesh 
while I like your simple approach, it come swith several drawbreaks that would break current behavoir as far as I can tell.

1. You cannot use hooks in the detail panel
2. You cannot respond to updated data since it will always only be rendered once
3. You break the rules of hooks by setting state on render, even though it might work limited in this use case
4. Rerenders are in themselves not bad and only if they are expensive something should be done to reduce them. As you can see, I logged them as well in my gif.

**Comment by appukuttan-shailesh** [comment]
> 1. You cannot use hooks in the detail panel

why is that?

> 2. You cannot respond to updated data since it will always only be rendered once

Not sure about this. I found that each time the panel expands, it is "mounted" and the functions executed and rendered. On collapsing, it is "unmounted". I believe it will show the updated data each time.

I certainly am more used to using class components than using functional components, so not really well informed about using hooks. 

> You break the rules of hooks by setting state on render, even though it might work limited in this use case

we could probably try to do this while setting the initial value itself (instead of null)?

> 4. Rerenders are in themselves not bad and only if they are expensive something should be done to reduce them.

My problem isn't with re-renders itself, but rather the detail panel's render might contain calls to functions (that can be expensive), and in this case they are being called unnecessarily a few times, rather than just on expand. This 'could potentially' be fixed by restructuring the user's code, but will break existing usage in several cases (my own use case being one of them, hence identified).

**Comment by appukuttan-shailesh** [comment]
p.s. I do propose to go ahead with #260 but with additional changes to avoid unnecessary renders.... also the unusual behavior I reported in my [other comment](https://github.com/material-table-core/core/pull/260#issuecomment-881406187).

PR #261 was made just as a demo of what behavior would be useful to have.

**Comment by Domino987** [comment]
Yes I a looking into that.

> You cannot use hooks in the detail panel
You are calling it in the function itself and not in the jsx context (here the return <>..</>). This prevents react to insert it into its notification system to reshedule udaptes.

> You cannot respond to updated data since it will always only be rendered once
Yes I mean while it is extended.

> we could probably try to do this while setting the initial value itself (instead of null)?
A useEffect would be the the best choice IMO.

What are you doing in your detail panel? Can you describe it so we might be able to adjust to it? Maybe a sandbox?

**Comment by appukuttan-shailesh** [comment]
I might need some time to make a sandbox of my present use case as many parts will need to be disconnected/replaced to reduce it. But we could proabably consider the following as an example use case:

https://codesandbox.io/s/material-collapse-table-forked-bgkbl?fontsize=14&hidenavigation=1&theme=dark&file=/demo.js

(above is just an example to demo an API call inside the detailpanel... the implementation isn't ideal/faultless)  

so basically involving some API calls.... but expensive ones.
So if we can update the detail panel such that the API calls (as in above example) get executed only once (i.e. only during each expand), that would be ideal I guess.... this would ensure that all current usages are left undisturbed.

**Comment by Domino987** [comment]
@appukuttan-shailesh So I updated the code to respect different detail panels and swapping between them.
There is a change of the syntax because we now render true react components instead of functions. This has several benefits. But to render we need to destruct the props.
`render: ({ rowData }) => {...}`
Rearding your code:
So I think you mistake rerender for remounts/calling of useEffects.

Since the useEffect has an empty dependency array ([]) as second parameter, it will not be called on the second render.
Just on true remounts, it would. But this is a different problem that should be adressed with caching.

**Comment by appukuttan-shailesh** [comment]
@Domino987 : Thanks for the update, it effectively resolves the bug demonstrated in the screenshot [above](https://github.com/material-table-core/core/pull/260#issuecomment-881406187).

> So I think you mistake rerender for remounts/calling of useEffects.

I am actually familar with the concept of renders vs mount/unmounts, though I might very likely be guilty of misusing the terms in our discussion ;-)

In #261 , I had actually initially toyed with useEffect() to handle component mount and unmount, but eventually seemed to find it was unnecessary for that particular implementation (as far as producing the desired output... you pointed out some other faults, which I wasn't aware of).

With regards to the current functionality, my only complaint is that it doesn't produce the same behavior as we had in v3.0.16. I added the same log statements to the demo app in v3.0.16:

Peek 2021-07-16 22-07

where the contained functions are only executed on expand..... i.e. in our case, the log statements would be executed once for each expand; and never on collpase (somewhat similiar to what I found with #261). I believe our intention is to retain the same behavior as v3.0.16 wrt detail panels, except for the sliding animation. This would ensure that there is no breakage for anyone.

Do you think it would be worth fixing the conceptual faults that you pointed out in #261, and seeing if we can retain the behavior of v3.0.16 in all other aspects. (I realize #261 isn't fit for acceptance, and was hoping you might be able to iron out the flaws)..... unless you already know how to fix this with the current PR itself. Thanks again for your work on this.

**Comment by matthewoestreich** [comment]
My fault on commenting on the wrong PR. I trust whatever @Domino987 chooses to do.
